File: research/data/metaforecast.py
"""metaforecast.org cross-platform fusion (P2b mirror).

Mirrors ``src/tools/finance/metaforecast.ts``. Pure helpers; no live HTTP.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Iterable

CROSS_PLATFORM_DELTA_THRESHOLD = 0.10

_STOPWORDS = frozenset({
    "will", "the", "a", "an", "be", "by", "in", "on", "of", "to", "and",
    "or", "is", "are", "was", "were", "has", "have", "had", "with", "for",
    "at", "as", "this", "that", "these", "those", "do", "does", "did",
})

_PUNCT_RE = re.compile(r"[^a-z0-9%\s]")


@dataclass(frozen=True)
class MetaforecastEstimate:
    title: str
    probability: float
    platform: str
    stars: int = 0
    url: str | None = None


def parse_metaforecast_response(raw: Any) -> list[MetaforecastEstimate]:
    """Parse a metaforecast `/api/v2/questions` payload."""
    if not isinstance(raw, list):
        return []
    out: list[MetaforecastEstimate] = []
    for r in raw:
        if not isinstance(r, dict):
            continue
        title = r.get("title")
        platform = r.get("platform")
        opts = r.get("options")
        if not isinstance(title, str) or not isinstance(platform, str):
            continue
        if not isinstance(opts, list) or not opts:
            continue
        prob: float | None = None
        for opt in opts:
            if not isinstance(opt, dict):
                continue
            try:
                p = float(opt.get("probability"))
            except (TypeError, ValueError):
                continue
            name = str(opt.get("name", "")).lower()
            if name == "yes" or "yes" in name:
                prob = p
                break
            if prob is None:
                prob = p
        if prob is None:
            continue
        clamped = max(0.0, min(1.0, prob))
        stars_raw = (r.get("qualityindicators") or {}).get("stars", 0)
        try:
            stars = int(stars_raw)
        except (TypeError, ValueError):
            stars = 0
        url = r.get("url") if isinstance(r.get("url"), str) else None
        out.append(MetaforecastEstimate(
            title=title,
            probability=clamped,
            platform=platform,
            stars=stars,
            url=url,
        ))
    return out


def _tokenize(text: str) -> set[str]:
    cleaned = _PUNCT_RE.sub(" ", text.lower())
    return {t for t in cleaned.split() if len(t) > 1 and t not in _STOPWORDS}


def find_best_metaforecast_match(
    question: str,
    candidates: Iterable[MetaforecastEstimate],
) -> MetaforecastEstimate | None:
    """Fuzzy keyword-overlap match (Jaccard ≥ 0.20)."""
    cand_list = list(candidates)
    if not cand_list:
        return None
    q_tokens = _tokenize(question)
    if not q_tokens:
        return None
    best: MetaforecastEstimate | None = None
    best_score = -1.0
    for c in cand_list:
        c_tokens = _tokenize(c.title)
        if not c_tokens:
            continue
        inter = len(q_tokens & c_tokens)
        union = len(q_tokens | c_tokens)
        jaccard = inter / union if union else 0.0
        if jaccard < 0.20:
            continue
        score = jaccard + c.stars * 1e-3
        if score > best_score:
            best_score = score
            best = c
    return best


def compute_cross_platform_delta(poly_prob: float, meta_prob: float) -> float:
    return abs(poly_prob - meta_prob)


def should_flag_cross_platform(delta: float) -> bool:
    return delta > CROSS_PLATFORM_DELTA_THRESHOLD


__all__ = [
    "MetaforecastEstimate",
    "parse_metaforecast_response",
    "find_best_metaforecast_match",
    "compute_cross_platform_delta",
    "should_flag_cross_platform",
    "CROSS_PLATFORM_DELTA_THRESHOLD",
]

File: src/tools/finance/metaforecast.ts
/**
 * P2b — metaforecast.org cross-platform fusion.
 *
 * Source: docs/polymarket-prediction-improvements-research-2026-07.md §6
 *
 * metaforecast.org aggregates Metaculus, Manifold, Kalshi, GJOpen, etc.
 * When Polymarket's price diverges from the cross-platform consensus by
 * more than ~10pp, that disagreement is a useful uncertainty signal —
 * fold it into MarketInput.warnings and apply a quality discount.
 *
 * This module exposes pure helpers (parsing + matching + scoring) and
 * a thin HTTP fetch that callers can opt into. The HTTP layer is not
 * unit-tested; mocks should target the pure helpers.
 */

const METAFORECAST_API = 'https://metaforecast.org/api/v2/questions';
const CROSS_PLATFORM_DELTA_THRESHOLD = 0.10;

export interface MetaforecastEstimate {
  /** Question title as published on the source platform. */
  title: string;
  /** YES-side probability, clamped to [0, 1]. */
  probability: number;
  /** Source platform (metaculus, manifold, kalshi, ...). */
  platform: string;
  /** Quality score (0 = unknown, 4 = highest). */
  stars: number;
  /** Optional URL back to the source. */
  url?: string;
}

/**
 * Parse a metaforecast `/api/v2/questions` payload into a typed list.
 * Tolerates missing fields and malformed entries (skips them).
 */
export function parseMetaforecastResponse(raw: unknown): MetaforecastEstimate[] {
  if (!Array.isArray(raw)) return [];
  const out: MetaforecastEstimate[] = [];
  for (const r of raw) {
    if (!r || typeof r !== 'object') continue;
    const rec = r as Record<string, any>;
    const title = typeof rec.title === 'string' ? rec.title : null;
    const platform = typeof rec.platform === 'string' ? rec.platform : null;
    const opts = Array.isArray(rec.options) ? rec.options : null;
    if (!title || !platform || !opts || opts.length === 0) continue;
    // Pick the YES probability — first option whose name looks affirmative,
    // else the first numeric one.
    let prob: number | null = null;
    for (const opt of opts) {
      if (!opt || typeof opt !== 'object') continue;
      const name = String(opt.name ?? '').toLowerCase();
      const p = Number(opt.probability);
      if (!Number.isFinite(p)) continue;
      if (name === 'yes' || name.includes('yes')) {
        prob = p;
        break;
      }
      if (prob === null) prob = p;
    }
    if (prob === null) continue;
    const clamped = Math.max(0, Math.min(1, prob));
    const stars = Number(rec.qualityindicators?.stars ?? 0);
    out.push({
      title,
      probability: clamped,
      platform,
      stars: Number.isFinite(stars) ? stars : 0,
      ...(typeof rec.url === 'string' ? { url: rec.url } : {}),
    });
  }
  return out;
}

const _STOPWORDS = new Set([
  'will', 'the', 'a', 'an', 'be', 'by', 'in', 'on', 'of', 'to', 'and',
  'or', 'is', 'are', 'was', 'were', 'has', 'have', 'had', 'with', 'for',
  'at', 'as', 'this', 'that', 'these', 'those', 'do', 'does', 'did',
]);

function _tokenize(text: string): Set<string> {
  return new Set(
    text
      .toLowerCase()
      .replace(/[^a-z0-9%\s]/g, ' ')
      .split(/\s+/)
      .filter((t) => t.length > 1 && !_STOPWORDS.has(t)),
  );
}

/**
 * Fuzzy keyword-overlap match. Returns the best candidate when the
 * Jaccard overlap is ≥ 0.20, else null. Ties broken by `stars`.
 */
export function findBestMetaforecastMatch(
  question: string,
  candidates: MetaforecastEstimate[],
): MetaforecastEstimate | null {
  if (!candidates.length) return null;
  const qTokens = _tokenize(question);
  if (qTokens.size === 0) return null;
  let best: MetaforecastEstimate | null = null;
  let bestScore = -1;
  for (const c of candidates) {
    const cTokens = _tokenize(c.title);
    if (cTokens.size === 0) continue;
    let inter = 0;
    for (const t of qTokens) if (cTokens.has(t)) inter += 1;
    const union = qTokens.size + cTokens.size - inter;
    const jaccard = union === 0 ? 0 : inter / union;
    if (jaccard < 0.20) continue;
    // Composite score: jaccard primary, stars tiebreaker.
    const score = jaccard + c.stars * 1e-3;
    if (score > bestScore) {
      bestScore = score;
      best = c;
    }
  }
  return best;
}

export function computeCrossPlatformDelta(polyProb: number, metaProb: number): number {
  return Math.abs(polyProb - metaProb);
}

export function shouldFlagCrossPlatform(delta: number): boolean {
  return delta > CROSS_PLATFORM_DELTA_THRESHOLD;
}

/**
 * Thin HTTP wrapper around metaforecast `/api/v2/questions`.
 * Network-dependent — call only from production code paths, not tests.
 */
export async function fetchMetaforecastQuestions(
  query: string,
  opts: { limit?: number; signal?: AbortSignal } = {},
): Promise<MetaforecastEstimate[]> {
  const limit = Math.max(1, Math.min(50, opts.limit ?? 10));
  const url = `${METAFORECAST_API}?query=${encodeURIComponent(query)}&limit=${limit}`;
  const res = await fetch(url, { signal: opts.signal });
  if (!res.ok) return [];
  const json = await res.json().catch(() => null);
  return parseMetaforecastResponse(json);
}
